overlord/servicestate/quota_handlers.go

```diff
@@ -123,54 +124,110 @@ func (m *ServiceManager) doQuotaControl(t *state.Task, _ *tomb.Tomb) error {
 		opts := &ensureSnapServicesForGroupOptions{
 			allGrps: allGrps,
 		}
-		appsToRestartBySnap, err = ensureSnapServicesForGroup(st, t, grp, opts)
+		servicesAffected, err = ensureSnapServicesForGroup(st, t, grp, opts)
 		if err != nil {
 			return err
 		}
 
 		// All persistent modifications to disk are made and the
 		// modifications to state will be committed by the
-		// unlocking in restartSnapServices. If snapd gets
+		// unlocking at the end of this task. If snapd gets
 		// restarted before the end of this task, all the
 		// modifications would be redone, and those
 		// non-idempotent parts of the task would fail.
 		// For this reason we record together with the changes
 		// in state the fact that the changes were made,
 		// to avoid repeating them.
 		// What remains for this task handler is just to
-		// restart services which will happen regardless if we
-		// get rebooted after unlocking the state - if we got
-		// rebooted before unlocking the state, none of the
-		// changes we made to state would be persisted and we
-		// would run through everything above here again, but
-		// the second time around EnsureSnapServices would end
-		// up doing nothing since it is idempotent.  So in the
-		// rare case that snapd gets restarted but is not a
+		// refresh security profiles and restart services which
+		// will happen regardless if we get rebooted after
+		// unlocking the state - if we got rebooted before unlocking
+		// the state, none of the changes we made to state would
+		// be persisted and we would run through everything above
+		// here again, but the second time around EnsureSnapServices
+		// would end up doing nothing since it is idempotent.
+		// So in the rare case that snapd gets restarted but is not a
 		// reboot also record which services do need
 		// restarting. There is a small chance that services
 		// will be restarted again but is preferable to the
 		// quota not applying to them.
-		if err := rememberQuotaStateUpdated(t, appsToRestartBySnap); err != nil {
+		if err := rememberQuotaStateUpdated(t, servicesAffected, refreshProfiles); err != nil {
 			return err
 		}
-
 	}
 
-	if err := restartSnapServices(st, t, appsToRestartBySnap, perfTimings); err != nil {
-		return err
+	if len(servicesAffected) > 0 {
+		prevTask := t
+		if refreshProfiles {
+			prevTask = addRefreshProfileTasks(st, prevTask, servicesAffected)
+		}
+		addRestartServicesTasks(st, prevTask, qc.QuotaName, servicesAffected)
```

Comment:
this code seems to assume that there are not more tasks after t waiting on it, that might be true right now, but is a bit of a dangerous assumption. 

See snapstate.InjectTasks for how we usually deal with this kind of issue

Reply:
I updated this to a different approach, I didn't do snapstate.InjectTasks as I would have to inject task-sets in reverse order which I didn't like. I dropped the 'prevTask' and made a different approach that instead uses TaskSets and we then add those task-sets to the current change and make sure they depend on each other that way. I feel like it's cleaner now than before, and would like to hear your opinion on this approach instead